pricegraph/src/orderbook/scalar.rs

```diff
@@ -169,3 +170,52 @@ fn assert_strictly_positive_and_finite(value: f64) -> f64 {
     debug_assert!(num::is_strictly_positive_and_finite(value));
     value
 }
+
+/// The exchange rate weight used by the pathfinding algorithm.
+#[derive(Clone, Copy, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Weight(i128);
+
+impl Weight {
+    /// Creates a new graph weight from a floating point number.
+    pub fn new(value: f64) -> Self {
+        // NOTE: We need to reserve some bits in order for the addition not to
+        // overflow, specifically we need to reserve
+        // - 16 bits for the maximum number of nodes (maximum number of tokens)
+        // - 8 bits for the [-128, 128) range
+        // - 1 bit for infinity
+        // Thats a total of 25 bits, which leaves us with 103 bits of precision
+        // for the weight.
+        const FACTOR: f64 = (1u128 << 103) as _;
+
+        let weight = value.log2() * FACTOR;
+        debug_assert!(-128.0 * FACTOR <= weight && weight < 128.0 * FACTOR);
+
+        Weight(weight as _)
+    }
+}
+
+impl FloatMeasure for Weight {
+    fn zero() -> Self {
+        Weight(0)
+    }
+
+    fn infinite() -> Self {
+        Weight(i128::MIN)
+    }
+}
+
+impl ops::Add for Weight {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        // NOTE: The Bellman-Ford implementation relies on `∞ * x == ∞`.
+        // TODO(nlordell): Since this branch is in a tight loop, we would
+        // probably benifit quite a bit in optimizing this so that there are no
+        // branches.
+        if self == Weight::infinite() || rhs == Weight::infinite() {
+            Weight::infinite()
+        } else {
+            Weight(self.0 + rhs.0)
+        }
```

Comment:
If we need to optimize this I'd do it in a follow up PR too unless it turns out that performance is terrible without it.
I don't think it will be super important because this is not going to compile to a branch (with jump) anyway. Instead the compiler should be able to use the cmove instruction: https://godbolt.org/z/G4sj3n
Still a lot more instructions for `i128` that can maybe be improved but it's more on the micro optimization level.

Reply:
> If we need to optimize this I'd do it in a follow up PR

Agree, that's why I left a todo.

> Instead the compiler should be able to use the cmove

Nice! I also compared the performance of this vs `saturated_add` and the if statement seems to perform slightly better and also panic on overflow in `debug` mode :metal:.

I'll reword the TODO.